### DISCUSSION TITLE
Extend npx create-payload-app with a frontend --template and/or --example

### DISCUSSION THREAD

**Discussion by 3m1l1a** [payloadcms/payload#1466] [Feature Requests & Ideas]
## Question
Where is the source code for create-payload-app?
I'm interested in extending the scaffolding tool to include frontend frameworks and themes.

### Next.js
[create-next-app](https://github.com/vercel/next.js/tree/canary/packages/create-next-app)

Next do have official examples for connecting a headless cms, but not Payload yet.
[official Next.js examples](https://github.com/vercel/next.js/tree/canary/examples)

**Payload's [nextjs-custom-server boilerplate](https://github.com/payloadcms/nextjs-custom-server)**

- [ ] [js-blank](https://vercel.com/templates/next.js/nextjs-boilerplate)
- [ ] js-blog
- [ ] js-todo
- [ ] [js-portfolio](https://vercel.com/templates/next.js/portfolio-starter-kit)
- [ ] ts-blank
- [ ] [ts-blog](https://vercel.com/templates/next.js/blog-starter-kit) **statically generated**
- [ ] ts-todo
- [ ] ts-portfolio

### Astro.build
[yarn create astro](https://docs.astro.build/en/install/auto/)
[Official Astro.build themes](https://astro.build/themes/official/)
[Astro guide Payload CMS](https://github.com/withastro/docs/blob/main/src/pages/en/guides/cms/payload.md)

[Lambdo Labs Payload CMS and Astro template](https://github.com/Lambdo-Labs/payloadcms-astro-template)

- [ ] [js-blank](https://github.com/withastro/astro/tree/main/examples/basics)
- [ ] [js-blog](https://github.com/withastro/astro/tree/main/examples/blog)
- [ ] js-todo
- [ ] [js-portfolio](https://github.com/withastro/astro/tree/main/examples/portfolio)
- [ ] ts-blank
- [ ] ts-blog
- [ ] ts-todo
- [ ] ts-portfolio

### Feature: extended help output with short flags and options
```
$ npx create-payload-app --help
Payload 1.2.0, the most powerful Typescript CMS.

Usage:
    
    Synopsis:
        $ npx create-payload-app[@Version] [Options]

    Example:
        Payload setup wizard:
        $ npx create-payload-app

        Create a named blank Payload version 1.2.0 and Nextjs app:
        $ npx create-payload-app@1.2.0 -n my-payload-app -t blank -f nextjs

Options:

    -d, --demo-content                  Also install demo content for specific template, framework and/or example
    
    -f, --framework                     Choose specific framework. Available frameworks: nextjs, astro
    
    -h, --help                          Show this help
    
    -l, --language                      Choose specific language. Available languages: javascript [js], typescript [ts]
      
    -n, --name my-payload-app           Set project name
      
    -e --example example-name           Choose specific white labeled example. Available examples: blank, blog, todo, portfolio

    -t --template template-name         Choose specific pre white labeled template. Available templates: blank, blog, todo, portfolio

    -V, --version                       Show create-payload-app version.
    
    -U, --use-npm                       Use npm to install dependencies

    -N, --no-deps                       Do not install any dependencies
```
### Feature: Scaffold a blank Payload and Nextjs typescript app

```
npx create-payload-app -n next-blank -t blank -f nextjs -l ts
```
  Welcome to Payload. Let's create a project!

✔ Choose project name › next-blank
✔ Choose project template › blank
✔ Choose project framework › nextjs
✔ Choose language › ts
✔ Enter MongoDB connection … mongodb://localhost/next-blank

  Creating a new Payload app in /home/3m1l1a/dev/create-payload-app/next-blank

✔ Project directory created
✔ Dependencies installed
✔ .env file created
✔ Payload project successfully created

  ★ Launch Application:
    - cd ./next-blank
    - `yarn dev` or `docker-compose up`

  ★ View Application:
    - Next.js: [http://localhost:3000]()
    - Payload: [http://localhost:3000/admin]()
    - Mongo Express: [http://localhost:8081]()
  
  ★ Read Documentation:
    - [Getting Started](https://payloadcms.com/docs/getting-started/what-is-payload)
    - [Configuration](https://payloadcms.com/docs/configuration/overview)

**Comment by denolfe** [comment]
All great suggestions @3m1l1a . 

We have work underway to get Payload into the Next.js examples repo. That PR should be opened shortly.

Regarding additional functionality, we've discussed internally about having a differentiation between starters and full demos. I like your idea of optionally adding a front-end. However, I have concerns of how maintainable that would be in the long-term.

In the meantime, we've open-sourced create-payload-app code here: https://github.com/payloadcms/create-payload-app.

**Comment by 3m1l1a** [comment]
Great to hear that Payload is on it's way into Next,js example repo.

I do share your concerns @denolfe . It might be difficult to maintain in the long-term, there is a lot of frameworks and themes. Payload should probably stay with one framework. Some sort of community plugin system would probably be better for the rest. My main idea with the front-end was to have something like WordPress has with their official yearly themes. Which would add this alternatives to create-payload-app:
- No front-end
- No front-end with your [demo](https://github.com/payloadcms/public-demo) 
- Official starter, kind of already exist here: [next-custom-server](https://github.com/payloadcms/nextjs-custom-server)
- Official full demo
 
Wonderful! I'will start scaffolding Astro's official themes and think more about a sustainable sulution.
